nexus/types/src/external_api/params.rs

```diff
@@ -18,34 +19,109 @@ use serde::{
 use std::{net::IpAddr, str::FromStr};
 use uuid::Uuid;
 
+#[derive(Deserialize, JsonSchema)]
+pub struct OrganizationPath {
+    pub organization: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct ProjectPath {
+    pub project: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct InstancePath {
+    pub instance: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct OrganizationSelector {
+    pub organization: NameOrId,
+}
+
+impl From<Name> for OrganizationSelector {
+    fn from(name: Name) -> Self {
+        OrganizationSelector { organization: name.into() }
+    }
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct OptionalOrganizationSelector {
+    #[serde(flatten)]
+    pub organization_selector: Option<OrganizationSelector>,
+}
```

Comment:
The only reason this optional selector has to exist is because dropshot chokes on `Query<Option<OrganizationalSelector>>`. If that worked we could delete this.

Reply:
The types actually work out fine, but the `EXPECTORATE` update fails. 

<details>
<summary>Example error</summary>

```
thread 'integration_tests::commands::test_nexus_openapi' panicked at 'expected normal process exit with code 0, got Signaled(6)

process stderr:thread 'main' panicked at 'invalid type Object(
    SchemaObject {
        metadata: Some(
            Metadata {
                id: None,
                title: Some(
                    "Nullable_ProjectSelector",
                ),
                description: None,
                default: None,
                deprecated: false,
                read_only: false,
                write_only: false,
                examples: [],
            },
        ),
        instance_type: None,
        format: None,
        enum_values: None,
        const_value: None,
        subschemas: Some(
            SubschemaValidation {
                all_of: Some(
                    [
                        Object(
                            SchemaObject {
                                metadata: None,
                                instance_type: None,
                                format: None,
                                enum_values: None,
                                const_value: None,
                                subschemas: None,
                                number: None,
                                string: None,
                                array: None,
                                object: None,
                                reference: Some(
                                    "#/components/schemas/ProjectSelector",
                                ),
                                extensions: {},
                            },
                        ),
                    ],
                ),
                any_of: None,
                one_of: None,
                not: None,
                if_schema: None,
                then_schema: None,
                else_schema: None,
            },
        ),
        number: None,
        string: None,
        array: None,
        object: None,
        reference: None,
        extensions: {
            "nullable": Bool(true),
        },
    },
)', /Users/just-be/.cargo/git/checkouts/dropshot-a4a923d29dccc492/2529126/dropshot/src/handler.rs:932:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
', nexus/tests/integration_tests/commands.rs:106:38
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

</details>

Reply:
I haven't tried converting all of it but you might be able to represent this as:

```rust
#[derive(Deserialize, JsonSchema)]
pub struct OrganizationSelector {
    pub organization: Option<NameOrId>,
}

#[derive(Deserialize, JsonSchema)]
pub struct ProjectSelector {
    #[serde(flatten)]
    pub organization_selector: Option<OrganizationSelector>,
    pub project: Option<NameOrId>,
}
```

It would mean that the error message in `_ => Err(Error::invalid_request(...)` would need updating.

Reply:
That's true. I made it more explicit because there are cases when we know the user must provide a path param (like the parent resource of a list endpoint) vs when it could be represented by the path. I think the extra code (which can be generated) is worth the trade off of a more correct spec.